src/Common/src/System/Collections/Generic/ArrayBuilder.cs

```diff
@@ -0,0 +1,216 @@
+    /// </summary>
+    internal struct ArrayBuilder<T>
+    {
+        private T[] _array; // Starts out null, initialized on first Add.
+        private int _count; // Number of items into _array we're using.
+
+        /// <summary>
+        /// Initializes the <see cref="ArrayBuilder{T}"/> with a specified capacity.
+        /// </summary>
+        /// <param name="capacity">The capacity of the array to allocate.</param>
+        public ArrayBuilder(int capacity) : this()
+        {
+            Debug.Assert(capacity >= 0);
+            if (capacity > 0)
+            {
+                _array = new T[capacity];
+            }
+        }
+
+        /// <summary>
+        /// Gets the number of items this instance can store without re-allocating,
+        /// or 0 if the backing array is <c>null</c>.
+        /// </summary>
+        public int Capacity => _array?.Length ?? 0;
+
+        /// <summary>
+        /// Gets the number of items in the array currently in use.
+        /// </summary>
+        public int Count => _count;
+
+        /// <summary>
+        /// Gets or sets the item at a certain index in the array.
+        /// </summary>
+        /// <param name="index">The index into the array.</param>
+        public T this[int index]
+        {
+            get
+            {
+                Debug.Assert(index >= 0 && index < _count);
+                return _array[index];
+            }
+            set
+            {
+                Debug.Assert(index >= 0 && index < _count);
+                _array[index] = value;
+            }
+        }
+
+        /// <summary>
+        /// Adds an item to the backing array, resizing it if necessary.
+        /// </summary>
+        /// <param name="item">The item to add.</param>
+        public void Add(T item)
+        {
+            if (_count == Capacity)
+            {
+                EnsureCapacity(_count + 1);
+            }
+
+            UncheckedAdd(item);
+        }
+
+        /// <summary>
+        /// Adds an array of items to the backing array, resizing it if necessary.
+        /// </summary>
+        /// <param name="items">The items to add.</param>
+        public void AddRange(T[] items)
+        {
+            Debug.Assert(items != null);
+
+            if (items.Length != 0)
+            {
+                int endCount = _count + items.Length;
+                Debug.Assert(endCount > 0); // Check for overflow
+
+                if (endCount > Capacity)
+                {
+                    EnsureCapacity(endCount);
+                }
+
+                Debug.Assert(Capacity > 0); // At least 1 item is being added per above branch
+                Array.Copy(items, 0, _array, _count, items.Length);
+                _count = endCount;
+            }
+        }
+
+        /// <summary>
+        /// Gets an enumerator which enumerates the contents of this builder.
+        /// </summary>
+        public Enumerator GetEnumerator()
```

Comment:
Where does it get enumerated?  Seems like the 99% use case is write-then-generate-array, and for the remaining cases, its indexer could be used.